# 숫자 문자열과 영단어

네오와 프로도가 숫자놀이를 하고 있습니다. 네오가 프로도에게 숫자를 건넬 때 일부 자릿수를 영단어로 바꾼 카드를 건네주면 프로도는 원래 숫자를 찾는 게임입니다.

다음은 숫자의 일부 자릿수를 영단어로 바꾸는 예시입니다.

1478 → "one4seveneight"

234567 → "23four5six7"

10203 → "1zerotwozero3"

이렇게 숫자의 일부 자릿수가 영단어로 바뀌어졌거나, 혹은 바뀌지 않고 그대로인 문자열 s가 매개변수로 주어집니다.

s가 의미하는 원래 숫자를 return 하도록 solution 함수를 완성해주세요.

참고로 각 숫자에 대응되는 영단어는 다음 표와 같습니다.

0	zero

1	one

2	two

3	three

4	four

5	five

6	six

7	seven

8	eight

9	nine

# 사고과정

1. 딕셔너리 구조같은데? + 리스트와 len으로도 가능하지 않을까?
2. 하나씩 바꿔야 하니 for 문으로
3. 그럼 변환하는 건 어떻게?? : replace 기능을 사용
4. 숫자열을 원하니 숫자열로 표기

In [17]:
# 풀이 1
s = "one4seveneight"

def solution(s):
    num_dict = {"zero": "0", "one": "1", "two": "2", "three": "3", "four": "4", "five": "5", "six": "6", "seven": "7", "eight": "8", "nine": "9"}

    for word, num in num_dict.items():   # 딕셔너리에서 요소를 하나씩 뽑아오기
        s = s.replace(word, num)         # replace("찾을단어", "바꿀단어")(+문자열만 됨)
    return int(s)                        # int 여야 문제가 풀림

print(solution(s))

1478


In [13]:
# 풀이 2
s = "one4seveneight"

def solution(s):
    nums = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']

    for i in range(len(nums)):           # nums의 길이(10개) 만큼 작업 진행
        s = s.replace(nums[i], str(i))   # nums 리스트에 있는 것들을 i = len(nums) 숫자로 바꿈(문자열)
    return int(s)                        # 문자열이니 숫자로 변경

print(solution(s))

1478


# Generator

한번에 한 결과를 만들어냄

""Return one Fibonacci element at a time ""

iterator 라는 객체 반환함

[] 대신 () 로 표현

In [213]:
# 제너레이터

def fib_gen(n):
  a,b=0,1
  while b<n :
    yield b               # b를 출력하고 다음 진행하고 여기부터 다시 진행
    a,b = b, a+b

for x in fib_gen(1000) :
  print(x, end=', ')

1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 

for x in fib_gen(1000): 이 시작되면 다음과 같은 일이 벌어집니다.

0. 초기화: a = 0, b = 1이 됩니다. b < 1000 조건이 맞으므로 루프 진입!

1. 첫 번째 루프: yield b를 만납니다. 현재 b인 1을 for문의 x로 던져주고 함수는 그 자리에 일시 정지합니다.

   print(x)가 실행되어 화면에 1이 찍힙니다.

2. 두 번째 루프: for문이 다음 값을 달라고 요청하면, 멈췄던 yield b 바로 다음 줄부터 시작합니다.

    a, b = b, a + b 가 실행되어 a = 1, b = 1이 됩니다.

   다시 while 조건 검사 후 yield b를 만나서 1을 던져주고 또 일시 정지!

3. 세 번째 루프: 다시 깨어나서 a, b = b, a + b를 합니다. a = 1, b = 2가 됩니다.

    yield b를 만나서 2를 던져주고 일시 정지!

4. 이후 반복: 이 과정을 거쳐 3, 5, 8, 13 ... 순으로 값이 하나씩 밖으로 나갑니다.

5. 종료: b가 1000보다 커지는 순간(여기서는 987 다음 숫자인 1597이 될 때) while b < n 조건이 깨지면서 함수가 자연스럽게 종료되고 for문도 끝납니다.

# 판다스 실습

In [33]:
# 실습 1 판다스 설명

import pandas as pd

data = {'state' : ['ohio', 'ohio', 'ohio', 'nevada', 'nevada'],             # 지역
        'year' : [2000, 2001,2002,2001,2002],                               # 연도
        'pop' : [1.5,1.7,3.6,2.4,2.9]}                                      # 인구
frame = pd.DataFrame(data)
frame2 = pd.DataFrame(data, columns=['year','state','pop','debt'],          # 컬럼 데이터
                            index = ['one', 'two', 'three', 'four', 'five'] # 인덱스 데이터
                      )

frame               # 그냥 만든 경우
frame2              # 조정한 경우
frame2.state        # (= frame2['state'])
frame2[0:3]         # frame2['one':'three']

frame2.loc['three'] # 속성값들 가져오기 # (= .iloc[2])
                    # 속성명 : loc / 위치값 : iloc

,year,state,pop,debt
one,2000,ohio,1.5,NaN
two,2001,ohio,1.7,NaN
three,2002,ohio,3.6,NaN
four,2001,nevada,2.4,NaN
five,2002,nevada,2.9,NaN


In [83]:
# 실습 1-1

frame2['state'] # state 만 출력                 # 시리즈 개체 형태

                # 또 다른 방법
                # frame2.iloc[:,1]
                # frame2.state

,state
one,ohio
two,ohio
three,ohio
four,nevada
five,nevada


In [216]:
# 실습 1-2

frame2[['state','pop','debt']] # 'state','pop','debt' 만 출력

# 다른 방법
# frame2.loc[:, ['state', 'pop', 'debt']]
# frame2.iloc[:, [1, 2, 3]]

,state,pop,debt
one,ohio,1.5,NaN
two,ohio,1.7,NaN
three,ohio,3.6,NaN
four,nevada,2.4,NaN
five,nevada,2.9,NaN


In [220]:
# 실습 1-3

frame2[1:2] # two 만 출력

# 다른 방법

# frame2.iloc[1:2]
# frame2.iloc[1:2, :]
# frame2.loc['two':'two']

,year,state,pop,debt
two,2001,ohio,1.7,NaN


In [48]:
# 실습 1-4

print(frame2[1:4]) # two~four 만 출력

       year   state  pop debt
two    2001    ohio  1.7  NaN
three  2002    ohio  3.6  NaN
four   2001  nevada  2.4  NaN


In [223]:
# 실습 1-5

frame2.loc[['three', 'four'], ['pop', 'debt']] # 일부만 출력

# frame2[['pop', 'debt']][2:4]                          # 인덱스값으로도 가능

# frame2.iloc[[2, 3], [2, 3]]
# frame2.iloc[2:4, 2:4]

,pop,debt
three,3.6,NaN
four,2.4,NaN


In [215]:
# 실습 1-6

print(frame2.loc[:, ['pop', 'debt', 'state']]) # 열 색인 정렬 순서 조정

print(frame2.iloc[:, [2, 3, 1]])               # 열 색인 정렬 순서 조정

       pop debt   state
one    1.5  NaN    ohio
two    1.7  NaN    ohio
three  3.6  NaN    ohio
four   2.4  NaN  nevada
five   2.9  NaN  nevada
       pop debt   state
one    1.5  NaN    ohio
two    1.7  NaN    ohio
three  3.6  NaN    ohio
four   2.4  NaN  nevada
five   2.9  NaN  nevada


In [214]:
# 실습 1-7

print(frame2.loc[['three', 'four', 'two'], :]) # 행 색인 정렬 순서 조정

print(frame2.iloc[[2, 3, 1], :])                # 행 색인 정렬 순서 조정

       year   state  pop debt
three  2002    ohio  3.6  NaN
four   2001  nevada  2.4  NaN
two    2001    ohio  1.7  NaN
       year   state  pop debt
three  2002    ohio  3.6  NaN
four   2001  nevada  2.4  NaN
two    2001    ohio  1.7  NaN


# 실습2 데이터의 형태

In [104]:
# 실습 2

frame2.values # 데이터는 nd.array 형태로 저장됨 # 데이터만 다루고 싶을 경우 활용

array([[2000, 'ohio', 1.5, nan],
       [2001, 'ohio', 1.7, nan],
       [2002, 'ohio', 3.6, nan],
       [2001, 'nevada', 2.4, nan],
       [2002, 'nevada', 2.9, nan]], dtype=object)

In [106]:
frame2.values[2:4,2:4]

array([[3.6, nan],
       [2.4, nan]], dtype=object)

In [107]:
frame2.isnull()

,year,state,pop,debt
one,False,False,False,True
two,False,False,False,True
three,False,False,False,True
four,False,False,False,True
five,False,False,False,True


In [109]:
frame2.isnull().values            # null 값인거

array([[False, False, False,  True],
       [False, False, False,  True],
       [False, False, False,  True],
       [False, False, False,  True],
       [False, False, False,  True]])

In [110]:
frame2.isnull().values.sum()       # 데이터 전체의 null 값 수

np.int64(5)

In [111]:
frame2.isnull().sum()              # 속성별 null 값 개수

,0
year,0
state,0
pop,0
debt,5


# 실습 3 파일 파싱

In [120]:
# 실습 3-1 pandas 파일 파싱 함수

ex1 = pd.read_csv('ex1.csv')
ex1

,a,b,c,d,message
0,1,2,3,4,hello
1,5,6,7,8,world
2,9,10,11,12,foo


In [124]:
# 실습 3-2 pandas 파일 파싱 함수

names = ['a','b','c','d','message']
                # 속성이 message 인걸 인덱스로 바뀜(컬럼 데이터가 사라짐)
ex2 = pd.read_csv('ex1.csv', index_col='message')
ex2

,a,b,c,d
message,,,,
hello,1,2,3,4
world,5,6,7,8
foo,9,10,11,12


In [122]:
ex1.values # 데이터 타입을 문자열로 인식 sum 불가능

array([[1, 2, 3, 4, 'hello'],
       [5, 6, 7, 8, 'world'],
       [9, 10, 11, 12, 'foo']], dtype=object)

In [125]:
ex2.values # 데이터 타입을 숫자열로 인식 sum가능

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12]])

In [ ]:
# ex2 = pd.read_csv('ex1.csv', + )

# + 에 적을 수 있는 것들


In [157]:
# 실습 3-3  멀티 인덱스                : index_col=['Key1','key2']

parsed = pd.read_csv('ex03_multiindex_csv.txt', index_col=['key1','key2'])
parsed

val1  val2
key1 key2            
one  a        1     2
     b        3     4
     c        5     6
     d        7     8
two  a        9    10
     b       11    12
     c       13    14
     d       15    16

In [158]:
# 실습 3-4 구분 문자가 공백일 경우     : sep='\s+'

parsed = pd.read_table('ex04_space.txt', sep='\\s+') # \ or r 을 붙여야함
parsed

,a,b,c
aaa,-0.12,-1.02,-0.61
bbb,0.90,0.33,-0.03
ccc,-0.26,-0.38,-0.21
ddd,-0.08,-0.30,1.10


In [159]:
# 실습 3-5 # 실습 3-5 예외행 건너뛰기            : skiprows=[해당 열]

parsed = pd.read_csv('ex05_skip.txt', skiprows = [0,2,3] )
parsed

,a,b,c,d,message
0,1,2,3,4,hello
1,5,6,7,8,world
2,9,10,11,12,foo


In [160]:
# 실습 3-6 특정 문자를 null 취급      : na_values = 딕셔너리 이름

sentinels = {'message':['foo','NA'], 'something':['two']}
parsed = pd.read_csv('ex06_NA.txt' , na_values = sentinels)
parsed

,something,a,b,c,d,message
0,one,1,2,3.0,4,NaN
1,NaN,5,6,NaN,8,world
2,three9,10,11,12.0,foo,NaN


# 실습 4 JSON 데이터

- null 값
- 리스트 마지막에 쉼표 있으면 안됨
- 객체(딕셔너리), 배열(리스트),문자열, 숫자, 불리언, null 로 구성
- 키값은 반드시 문자열

# 실습 5 PICKLE 데이터

- 컴퓨터가 가장 빠르게 읽을 수 있는 이진 데이터 형식
- 실제로 보면 깨진 것처럼 보임

In [126]:
# 마운트하기

from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [137]:
# 실습 6 그냥 작업환경에 업로드 한 경우

parsed = pd.read_csv('ex03_multiindex_csv.txt', index_col=['key1','key2'])
parsed

val1  val2
key1 key2            
one  a        1     2
     b        3     4
     c        5     6
     d        7     8
two  a        9    10
     b       11    12
     c       13    14
     d       15    16

In [142]:
# 실습 6 구글 드라이브 마운트해서 작업 디렉토리 변경하여 작업

import os
print(os.getcwd())  # 작업 디렉토리 변경

os.chdir('./gdrive/MyDrive/Colab Notebooks/05 판다스 넘파이 실습/Day_01_file')
os.getcwd()

/content


'/content/gdrive/MyDrive/Colab Notebooks/05 판다스 넘파이 실습/Day_01_file'

In [204]:
# 종합 실습 # 1 # county, bush, gore 출력

#col_names = ["county", "technology", "columns", "under", "over", "Bush", "Gore", "Browne", "Nader", "Harris", "Hagelin", "Buchanan", "McReynolds", "Phillips", "Moorehead", "Chote", "McCarthy"]
fl2000 = pd.read_csv('fl2000.txt' ,
                     skiprows = 42,      # 무시 행 적용
                     sep='\\s+',         # 공백으로 구분된 거 처리
                     index_col= 'county' # 인덱스 적용
                     )
fl2000[['Bush', 'Gore']]

,Bush,Gore
county,,
Alachua,34124,47365
Baker,5610,2392
Bay,38637,18850
Bradford,5414,3075
Brevard,115185,97318
...,...,...
Union,2332,1407
Volusia,82357,97304
Wakulla,4512,3838


In [211]:
# 종합 실습 #2 # county, bush, gore 출력

#col_names = ["county", "technology", "columns", "under", "over", "Bush", "Gore", "Browne", "Nader", "Harris", "Hagelin", "Buchanan", "McReynolds", "Phillips", "Moorehead", "Chote", "McCarthy"]
fl2000 = pd.read_csv('fl2000.txt' ,
                     skiprows = range(42), # 무시 행 적용
                     delimiter='\\s+',     # 공백으로 구분된 거 처리
                     index_col= 'county'   # 인덱스 적용
                     )

fl2000.iloc[:,4:6]                         # 확인할 컬럼 적용

,Bush,Gore
county,,
Alachua,34124,47365
Baker,5610,2392
Bay,38637,18850
Bradford,5414,3075
Brevard,115185,97318
...,...,...
Union,2332,1407
Volusia,82357,97304
Wakulla,4512,3838
